vscode_extension/src/AacLanguageServer.ts

```diff
@@ -0,0 +1,104 @@
+import * as fs from "fs";
+import { ExtensionContext, workspace } from "vscode";
+import { LanguageClient, LanguageClientOptions, ServerOptions, Trace } from "vscode-languageclient/node";
+import { ensureTrue, execShell, getConfigurationItem, showMessageOnError } from "./helpers";
+
+const MIN_REQUIRED_PYTHON_VERSION = "3.9";
+
+export class AacLanguageServerClient {
+    private static instance: AacLanguageServerClient;
+
+    private aacLspClient!: LanguageClient;
+
+    private constructor() { }
+
+    public static getLspClient(): AacLanguageServerClient {
+        if (!AacLanguageServerClient.instance) {
+            AacLanguageServerClient.instance = new AacLanguageServerClient();
+        }
+        return AacLanguageServerClient.instance;
+    }
+
+    public startLanguageServer(context: ExtensionContext): void {
+        this.ensureAacToolIsAvailable();
+        this.ensureLspServerIsReady(context);
+    }
+
+    public shutdownServer(): void {
+        this.aacLspClient.stop();
+    }
+
+    private ensureAacToolIsAvailable(): void {
+        const pythonPath = this.getConfigurationItemFile("pythonPath");
+        this.ensureCorrectPythonVersionIsInstalled(pythonPath);
+        this.ensureCorrectAacVersionIsInstalled(pythonPath);
+    }
+
+    private getConfigurationItemFile(name: string): string {
+        const item: string = getConfigurationItem(name);
+        ensureTrue(item.length > 0, `Cannot start Language Server; '${item}' is not configured!`);
+        ensureTrue(fs.existsSync(item), `Cannot use ${item} as it does not exist!`);
+        return item;
+    }
+
+    private async ensureCorrectPythonVersionIsInstalled(pythonPath: string): Promise<void> {
+        const resolve = await execShell(`${pythonPath} --version`, {});
+        ensureTrue(!resolve.stderr, `Could not get the Python version.\n${resolve.stderr}`);
+
+        const pythonVersion = resolve.stdout.match(/\d+\.\d+\.\d+/)?.pop() ?? "unknown";
+        ensureTrue(
+            pythonVersion.startsWith(MIN_REQUIRED_PYTHON_VERSION),
+            `The AaC tool requires Python ${MIN_REQUIRED_PYTHON_VERSION} or newer; current version is: ${pythonVersion}`,
+        );
+    }
+
+    private async ensureCorrectAacVersionIsInstalled(pythonPath: string): Promise<void> {
+        const resolve = await execShell(`${pythonPath} -m pip freeze`);
+        ensureTrue(!resolve.stderr, `Could not get the installed AaC version.\n${resolve.stderr}`);
+
+        const expectedAacVersion: string = getConfigurationItem("version") ?? "";
+        const aacVersionPattern = new RegExp(`aac==${expectedAacVersion}|-e git.*egg=aac`, "ig");
+        const actualAacVersion = resolve.stdout.match(aacVersionPattern)?.pop();
+        ensureTrue(!!actualAacVersion, `The installed aac version is ${actualAacVersion} which is unsupported.`);
```

Comment:
I tried to find if the `!!` has another meaning than `not not`, but I couldn't find any resources that indicated that this isn't equivalent to:
```suggestion
        ensureTrue(actualAacVersion, `The installed aac version is ${actualAacVersion} which is unsupported.`);
```

I don't believe this is covered by a test, and I couldn't get the behavior to trigger when trying to run it. So, let me know if I'm missing something here.

Reply:
This is a JavaScript convention used when you want the [truthiness of a value](https://stackoverflow.com/a/784946/2247345). Since we're using TypeScript and `ensureTrue` takes a `bool` as the first parameter I needed to convert this to a `bool` value.